### PR TITLE
🎉metagpt tool - garden step

### PR DESCRIPTION
**Pull request by veronikasamborska1994**
hey @lucasrodes! no rush to take a look at this but it would be great if you could add the necessary bits of code so that garden metadata files start to show up in wizard as well! I am not sure where that step is so it's probably faster for you to do this. I really like the wizard interface and it would make it easier to play around with the garden tool to see where it fails :) 

If this ends up working well I think I might restructure the grapher step to do the metadata updates similarly (on an indicator level rather than the dataset level) as I think it will be less prone to running into text length limit issues!

### CODE REVIEW

**Comment by veronikasamborska1994** [comment]
> Thanks for working on this, @veronikasamborska1994!
> 
> I have added support for GARDEN files in `etl-wizard`. I was testing its behaviour for garden step `data://garden/war/2023-09-21/ucdp_prio.meta.yml` and it looks like there are several queries to GPT. Is that expected? I got this logging in the terminal:
> 
> ```
> 2023-12-29 13:09:49 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:09:51 [info     ] Cost GPT4: $0.03
> 2023-12-29 13:09:52 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:09:52 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:09:54 [info     ] Cost GPT4: $0.03
> 2023-12-29 13:09:55 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:09:56 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:09:58 [info     ] Cost GPT4: $0.03
> 2023-12-29 13:09:59 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:09:59 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:01 [info     ] Cost GPT4: $0.03
> 2023-12-29 13:10:02 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:03 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:05 [info     ] Cost GPT4: $0.03
> 2023-12-29 13:10:06 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:07 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:09 [info     ] Cost GPT4: $0.03
> 2023-12-29 13:10:12 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:12 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:15 [info     ] Cost GPT4: $0.03
> 2023-12-29 13:10:16 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:17 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:19 [info     ] Cost GPT4: $0.03
> 2023-12-29 13:10:21 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:22 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:24 [info     ] Cost GPT4: $0.03
> 2023-12-29 13:10:25 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:25 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:27 [info     ] Cost GPT4: $0.03
> 2023-12-29 13:10:27 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:28 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:30 [info     ] Cost GPT4: $0.03
> 2023-12-29 13:10:31 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:32 [info     ] Cost GPT4: $0.02
> 2023-12-29 13:10:33 [info     ] Cost GPT4: $0.03
> 2023-12-29 13:10:34 [info     ] Cost GPT4: $0.02
> ```

thanks a lot @lucasrodes! yep, because I am doing it on an indicator level in the garden step so it will make a query for each indicator (and also actually each metadata field so that we def don't run into length issues)! I will change this to output the final cost though!

**Comment by lucasrodes** [comment]
@veronikasamborska1994 I've added some improvements to the code base:

- Backend code
  - I've re-organinzed the code in `apps/metagpt`. It now consists of multiple modules:
    - `cli`: main module, with client code.
    - `gpt`: module with wrappers and classes to ease interaction with OpenAI API.
    - `prompts`: contains prompts for gpt, functions to generate queries for gpt, etc.
    - `utils`: generic utils to be used throughout the other modules.
- Wizard code:
  - Some minor improvements on navigation
  - Cost estimates are now shown prior to querying OpenAI. Needs user approval. We could think of setting a threshold in the future to run this if below X USD.


I am leaving tomorrow afternoon for ~1 week, so I'd suggest we merge this soon. Improvements (e.g. cost threshold, snapshot cost estimation, etc.) can be done in a new and separate PR.

File: apps/metagpt/cli.py [hunk 1]
```diff
@@ -128,8 +55,18 @@ def main(path_to_file: str, output_dir: str, overwrite: bool) -> None:
...
+            proceed = input(f"The total estimated cost is ${cost_estimated:.3f}. Do you want to proceed? (yes/no): ")
+            if proceed.lower() == "yes":
+                # Actually run
+                final_cost = gpt_updater.run()
+                log.info(f"Cost GPT4: ${final_cost:.3f}")
+        else:
+            final_cost = gpt_updater.run()
+            log.info(f"Cost GPT4: ${final_cost:.3f}")
         # Save updated metadata
         gpt_updater.save_updated_metadata(output_file_path)
```
Comment (veronikasamborska1994) [inline apps/metagpt/cli.py:71]: it seems like in wizard the `gpt_updater.save_updated_metadata(output_file_path)` line never gets executed for me. I've been playing around with using the custom-written `yaml_dump` function but I don't think it actually changes anything and if I uncomment/log anything here it won't throw an error/log in the terminal. I am probably being super stupid here but I don't understand how it currently saves the updated yaml without executing this step.

File: apps/metagpt/cli.py [hunk 2]
```diff
@@ -181,200 +118,134 @@ def metadata_new_str(self: Self) -> str:
...
+            log.error(f"Required garden dataset {self.path_to_dataset} does not exist. Run the garden step first.")
+            raise Exception("Required garden dataset does not exist. Run the garden step first.")
+
+    def load_yaml_metadata(self: Self) -> Dict[str, Any]:
+        """Load dataset metadata from YAML."""
+        with open(self.path_to_file, "r") as file:
+            original_yaml_content = yaml.safe_load(file)
+        return original_yaml_content
+
     def save_updated_metadata(self: Self, output_file: str) -> None:
```
Comment (veronikasamborska1994) [inline apps/metagpt/cli.py:155]: in wizard this function never gets called but it works ok now from the terminal

File: apps/wizard/metagpt.py [hunk 3]
```diff
@@ -31,9 +31,27 @@ def export_metadata_file(metadata_new, overwrite, output_dir, filepath) -> None:
         file.write(metadata_new)
```
Comment (veronikasamborska1994) [inline apps/wizard/metagpt.py:31]: that's a great log message haha :D I think we should use the yaml_dump here though to output everything nicely!  ie. --> ` file.write(yaml_dump(metadata_new))`

File: apps/wizard/metagpt.py [hunk 4]
```diff
@@ -134,12 +190,25 @@ def get_actual_path(path: str) -> str:  # -> Any:
...
-            overwrite = st.toggle("Overwrite", value=False)
-            st.form_submit_button(
-                "Export new file",
-                on_click=export_metadata_file,
-                args=(metadata_new_updated, overwrite, output_dir, filepath),
-            )
+        with st.expander("Export metadata file", expanded=False):
+            # Form to export
+            with st.form("form_export"):
+                overwrite = st.toggle(
```
Comment (veronikasamborska1994) [inline apps/wizard/metagpt.py:196]: this line isn't working as expected - still just append the gpt_ to the filename and saves it in the directory

File: apps/metagpt/gpt.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                **kwargs,
+            }
+        chat_completion = self.chat.completions.create(**kwargs)  # type: ignore
+        # Return value only if message content is not None
+        if isinstance(chat_completion, ChatCompletion) and isinstance(chat_completion.usage, CompletionUsage):
+            chat_completion_tokens = chat_completion.usage.total_tokens
+            cost = (chat_completion_tokens / 1000) * RATE_PER_1000_TOKENS
+            message_content = chat_completion.choices[0].message.content
+            # Log cost
+            log.info(f"Chat completion {self.model} cost: {cost}")
```
Comment (veronikasamborska1994) [inline apps/metagpt/gpt.py:99]: we probably don't need to log the cost at each step but what we could do is implement a progress bar for garden/grapher step to show where we are in the loop of indicators+ metadata field updates

Comment (veronikasamborska1994) [reply apps/metagpt/gpt.py:99]: I can easily make it work in the terminal but not sure how to do it in wizard afterwards...

File: apps/metagpt/cli.py [hunk 6]
```diff
@@ -112,7 +40,6 @@
 def main(path_to_file: str, output_dir: str, overwrite: bool) -> None:
     """Process and update metadata using GPT-based tool.
 
-    If `path-to-file` must either be a 'snapshot' or a 'grapher' metadata file.
     To learn more about the behaviour of this tool, please refer to https://docs.owid.io/projects/etl/architecture/metadata/.
     """
     log.info("Starting metadata update process.")
```
[the remark this answers is not in the document]
Comment (lucasrodes) [reply apps/metagpt/cli.py:45]: I will move this comment to the other thread you opened targeting the wizard file.

File: apps/wizard/metagpt.py (hunk 4 above)
Comment (lucasrodes) [reply apps/wizard/metagpt.py:196]: Copy-pasting from Veronika's comment:

> the --overwrite property isn't working as expected in wizard, and even after pressing overwrite I am pretty sure it still executes this line  output_file_path = os.path.join(output_dir, "gpt_" + os.path.basename(path_to_file)). I am not sure how this is linked in wizard so maybe you could take a look?

File: apps/wizard/metagpt.py (hunk 3 above)
Comment (lucasrodes) [reply apps/wizard/metagpt.py:31]: oh wow, lol ー I'll change the wording here :D.

I've updated the code, and it renders way nicer now! Thanks for the catch.

File: apps/metagpt/cli.py (hunk 1 above)
Comment (lucasrodes) [reply apps/metagpt/cli.py:71]: do you mean in `wizard`? Or also happening to the command client?

File: apps/metagpt/cli.py (hunk 2 above)
Comment (lucasrodes) [reply apps/metagpt/cli.py:155]: Saving the file differently on wizard. Probably the code in wizard is unnecessary complex, but that's the way I found to allow the user to further edit the GPT returned content.

File: apps/wizard/metagpt.py (hunk 4 above)
Comment (lucasrodes) [reply apps/wizard/metagpt.py:196]: I think this is now solved

File: apps/metagpt/gpt.py (hunk 5 above)
Comment (lucasrodes) [reply apps/metagpt/gpt.py:99]: we can do this in a future PR

File: apps/metagpt/cli.py (hunk 1 above)
Comment (veronikasamborska1994) [reply apps/metagpt/cli.py:71]: yep, only in wizard! I think it's all good now though! happy to merge this one and then I'll create a new branch to play around with adapting the prompt messages

**Pull request merged**